internal/discovery/discovery.go

```diff
@@ -735,13 +731,62 @@ func (d *Discovery) processFile(path string, l log.Logger, filenames []string) c
 
 	canonicalDir, canErr := util.CanonicalPath(dir, d.workingDir)
 	if canErr == nil {
-		for _, pattern := range d.compiledExcludePatterns {
-			if pattern.Compiled.Match(canonicalDir) {
-				l.Debugf("Path %s excluded by glob %s", canonicalDir, pattern.Original)
+		// Eventually, this is going to be removed entirely, as filter evaluation
+		// will be all that's needed.
+		if !d.filterFlagEnabled {
+			for _, pattern := range d.compiledExcludePatterns {
+				if pattern.Compiled.Match(canonicalDir) {
+					l.Debugf("Path %s excluded by glob %s", canonicalDir, pattern.Original)
+					return nil
+				}
+			}
+		}
+
+		if d.filterFlagEnabled {
+			cfg := d.createComponentFromPath(path, filenames)
+			if cfg == nil {
 				return nil
 			}
+
+			// Check for hidden directories before returning
+			if !d.hidden && d.isInHiddenDirectory(path) {
+				allowHidden := false
+
+				// Always allow .terragrunt-stack contents
+				cleanDir := util.CleanPath(canonicalDir)
+				for part := range strings.SplitSeq(cleanDir, "/") {
+					if part == config.StackDir {
+						allowHidden = true
```

Comment:
I think this code can be simplified into separate function called `isInStackDirectory(path)`